### PR TITLE
Fix incorrect BitmapText text width calculation (#7758)

### PR DESCRIPTION
**Pull request by kjarrio**
##### Description of change
Fix for BitmapText width calculation when using monospace fonts. Fixes issue #7758 - and doesn't break PR #7479 / #7245

##### Pre-Merge Checklist
- [x] Lint process passed (`npm run lint`)
- [x] Tests passed (`npm run test`)

### CODE REVIEW

**Comment by kjarrio** [comment]
Example: https://codesandbox.io/s/festive-cartwright-7ruly?file=/index.html

**Comment by bigtimebuddy** [comment]
Thank you for the PR! Could you possibly add unit test for this behavior?

**Review by neymanushka — APPROVED** [review]

File: packages/text-bitmap/src/BitmapText.ts [hunk 1]
```diff
@@ -346,10 +356,12 @@ export class BitmapText extends Container
                 pos.y += data.lineHeight;
                 prevCharCode = null;
                 spaceCount = 0;
+                lastLineWidth = lastBreakWidth;
             }
         }
 
         const lastChar = charsInput[charsInput.length - 1];
+        lastLineWidth += Math.abs(xOffsetWidthComp);
```
Comment (bigtimebuddy) [inline packages/text-bitmap/src/BitmapText.ts:364]:
```suggestion

        lastLineWidth += Math.abs(xOffsetWidthComp);
```

**Comment by kjarrio** [comment]
Here is work in progress, for proper width/height calculations for BitmapTexts:

Before: https://63hr2.csb.app PixiJS 6.2.0 (Source: https://codesandbox.io/s/boring-shadow-63hr2?file=/index.html )
After: https://wiwen.csb.app (Source: https://codesandbox.io/s/tender-violet-wiwen?file=/index.html )

I have the unit tests as well, but I'm still working on a few edge cases for "pixel perfection" but I'm unsure if it's even worth it. For me, it's "good enough" with those changes. How important is it to get this 100%?

I'll tidy up the code to finish this PR as soon as I have time.

**Comment by bigtimebuddy** [comment]
Sorry it's taken a long time to review!

Seeing two issues here with width in your recent examples:

"C" is hanging off:
Screen Shot 2021-12-13 at 1 49 00 PM

Character endings are peaking out:
Screen Shot 2021-12-13 at 1 49 05 PM

**Comment by kozickikarol** [comment]
@kjarrio Are you going to fix something here? I would gladly use new version with your fix cause I encounter the same problem.

**Review by bigtimebuddy — APPROVED** [review]
This isn't perfect, but it's good enough. If we can improve it overtime that'd be great.

**Comment by kozickikarol** [comment]
> This isn't perfect, but it's good enough. If we can improve it overtime that'd be great.

🎉 
@bigtimebuddy Any chance to have it in upcoming 6.2.2? It blocks our project from migrating to v6.

**Comment by kozickikarol** [comment]
@kjarrio I tested it more on my monospaced fonts and I noticed that this solution does not work for fonts which has xadvance<width or negative xoffset. BitmapText behaves correctly when I change digits inside but when first or last digit changes, BitmapText starts shaking.

monospaced font with negative xoffset:
[monospace_font.zip](https://github.com/pixijs/pixijs/files/7897559/monospace_font.zip)

Same font works correctly in pixi v5.

**Comment by joel-Yolted** [comment]
as @kozickikarol pointed this fix does not solve the issue, There is still a clear regression made by  the original #7245 fix breaking xAdvance and incorrectly calculates the lastLineWidth of the character when xAdvance is used as shown with the original recreation on #7758 using 3.6.0  https://jsfiddle.net/1zyrd4s8/
It seems to be only occurring when anchor is set to 0.5 but that is a common use case.

Was this an intentional break? and if so please advice on why.

**Review by bigtimebuddy — CHANGES_REQUESTED** [review]
This needs more work based on the feedback here with negative offsets.

**Comment by bigtimebuddy** [comment]
Ping @kjarrio do you want to continue working on this? If not I'd like to close and maybe someone else can pick-up this work.

**Comment by kjarrio** [comment]
I will work on this, starting today, I can't promise I'll be able to fix it, but I will try to get this perfect this week

**Comment by Zyie** [comment]
Hey,

Thank you for putting together this PR! Since we've moved to the next major version of the library, we're closing this to focus on the updated codebase. If this is still relevant, feel free to submit a new PR for the latest version.
